sketch/src/sketch_rnn/model.ts

```diff
@@ -0,0 +1,363 @@
+    this.initialized = false;
+  }
+
+  /**
+   * Returns true if model is intialized.
+   */
+  isInitialized() {
+    return this.initialized;
+  }
+
+  /**
+   * Instantiates class information
+   * inputs from the JSON model.
+   */
+  private instantiateFromJSON(info: SketchRNNInfo,
+    weightDims: number[][],
+    weightStrings: string[]) {
+
+    this.forgetBias = tf.scalar(1.0);
+    this.info = info;
+    this.set_pixel_factor(2.0);
+    this.weightDims = weightDims;
+    this.numUnits = this.weightDims[0][0]; // size of LSTM
+    let rawWeights: Float32Array;
+    const maxWeight = 10.0;
+    this.weights = [];
+    for (let i = 0; i < weightStrings.length; i++) {
+      rawWeights = new Float32Array(support.string_to_array(weightStrings[i]));
+      const N: number = rawWeights.length;
+      for (let j = 0; j < N; j++) {
+        rawWeights[j] = maxWeight * rawWeights[j] / 32767;
+      }
+      this.weights.push(rawWeights);
+    }
+    this.outputKernel = tf.tensor2d(this.weights[0],
+      [this.weightDims[0][0], this.weightDims[0][1]]);
+    this.outputBias = tf.tensor1d(this.weights[1]);
+    const lstmKernelXH = tf.tensor2d(this.weights[2],
+      [this.weightDims[2][0], this.weightDims[2][1]]);
+    const lstmKernelHH = tf.tensor2d(this.weights[3],
+      [this.weightDims[3][0], this.weightDims[3][1]]);
+    const axis = 0;
+    this.lstmKernel = tf.concat2d([lstmKernelXH, lstmKernelHH], axis);
+    this.lstmBias = tf.tensor1d(this.weights[4]);
+
+    this.rawVars = [
+      this.outputKernel,
+      this.outputBias,
+      this.lstmKernel,
+      this.lstmBias
+    ];
+
+  }
+
+  /**
+   * Loads variables from the JSON model
+   */
+  async initialize() {
+    this.dispose();
+
+    const vars = await fetch(this.checkpointURL)
+      .then((response) => response.json());
+
+    this.instantiateFromJSON(vars[0], vars[1], vars[2]);
+
+    this.initialized = true;
+    console.log('Initialized SketchRNN.');
+  }
+
+  dispose() {
+    if (this.rawVars) {
+      for (let i = 0; i < this.rawVars.length; i++) {
+        this.rawVars[i].dispose();
+      }
+      this.rawVars = undefined;
+    }
+    if (this.forgetBias) {
+      this.forgetBias.dispose();
+      this.forgetBias = undefined;
+    }
+    this.initialized = false;
+  }
+
+  /**
+   * sets the internal EXTRA factor of this model (pixel to model space)
+   *
+   * @param scale (the extra scale factor for pixel to model space)
+   *
+   * @returns nothing
+   */
+  set_pixel_factor(scale: number) {
+    // for best effect, set to 1.0 for d3 or paper.js, 2.0 for p5.js
+    this.pixelFactor = scale;
+    this.scaleFactor = this.info.scale_factor / this.pixelFactor;
+  }
+
+  /**
+   * Updates the RNN, returns the next state.
+   *
+   * @param pen [dx, dy, pen_down, pen_up, pen_end].
+   * @param state previous [c, h] state of the LSTM.
+   *
+   * @returns the next [c, h] state of the LSTM`.
+   */
+  update(pen: number[], state: Float32Array[]) {
```

Comment:
I've thought hard about this, and it was a design choice is I need to deal with the state outside of the TensorFlow.js library. Your suggestion is great and in the future I can write the functionality to predict a set of N strokes, and also encode the previous M strokes.

Below are the reasons why I really need to do the update outside of the model, at the expense of some performance:

1) If an interactive artist wants to put real time knobs to adjust the temperature *during* the sketch generation in a performance setting, or even change classes, it is more straight forward to do if we store the state outside and have it generate one small stroke at a time.

2) Flexibility to generate art that is not exactly the QuickDraw domain. If an artist wants to generate images that are not just QuickDraw, but insert his/her handcrafted aesthetic augmentations into the code (for instance, change the maximum angle possible at each small line segment to ensure a smooth aesthetic), these things are easier to experiment on the "client side", giving us the ability to modify each line segment proposed by sketch-rnn, before feeding it back.

3) The artist can use the current scheme to easily extend to other purposes, like generating 16 sketches in parallel, if that is something he/she wants to do. Of course it is possible to do this also by adding functions to the main library to take advantage of batch processing, but for one-off installations, we might not be able to think of all possible use-cases, and the artist who is familiar with p5.js might not be able to hack the TensorFlow.js typescript-based code to add custom features easily, so having a possible interface to do in raw js in p5.js at the expense of some performance is worthwhile

4) When profiling the code, the main bottleneck of sketch-rnn is the one large matmul of the LSTM. Doing everything in tensor-space will save time to avoid the dataSyncs, but for the applications I've tested, the dataSync time is minimal compared to the large LSTM matmul which provides the bulk of the performance improvement.

I think leaving the current method in, and adding in methods to do multi-step generation, and multi-step encoding for performance will be the best of both worlds.

Reply:
Agreed. Would you mind adding these "batch" methods?

Reply:
Sounds good. I'll put in the multi-step update method (since it's going to be useful now for current applications), but due to some time constraints on my end to work on another project, I'll do the multi-step generation method in the future (as a TODO), since current interactive apps don't need it yet (but future applications might).